pageserver/src/virtual_file/owned_buffers_io/write.rs

```diff
@@ -107,31 +109,59 @@ where
     }
 
     #[cfg_attr(target_os = "macos", allow(dead_code))]
-    pub async fn flush_and_into_inner(
-        mut self,
-        ctx: &RequestContext,
-    ) -> std::io::Result<(u64, Arc<W>)> {
-        self.flush(ctx).await?;
+    pub async fn shutdown(mut self, ctx: &RequestContext) -> std::io::Result<(u64, W)> {
+        let buf = self.tail_mut();
+        let len = buf.pending();
+        let cap = buf.cap();
+        if len < cap {
+            // pad zeros to the next io alignment requirement.
+            let count = len.next_multiple_of(B::ALIGN).min(cap) - len;
+            buf.extend_with(0, count);
+        }
```

Comment:
Direct IO alignment requirement:

- buffer at aligned memory address.
- file offset to write at is aligned.
- buffer size written out is multiple of alignment.

Without the padding, `VirtualFile::write_all_at` fails with `invalid argument` because the buffer size we are writing out is potentially not a multiple of 512.

Reply:
38f1150ae29bda4952c52e0bdae9346e21e78a69